cmd/clusterctl/pkg/client/config/client.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/internal/test"
+)
+
+// Client is used to interact with the clusterctl configurations.
+// Clusterctl v2 handle two types of configs:
+// 1. The configuration of the providers (name, type and URL of the provider repository)
+// 2. Variables used when installing providers/creating clusters. Variables can be read from the environment or from the config file
+type Client interface {
```

Comment:
It's not overly clear to me why we need a single `Client` abstracting both the `ProvidersClient` and the `VariablesClient` here other than to dedup the small amount of code in `newConfigClient`. I think I'd prefer to see us avoid tying to much of the exposed API to require someone initialize both types of clients in favor of the type of client they are choosing to consume.

Reply:
@detiber the main reason to have a config client grouping `ProviderClient` and `VariablesClient` is to create an internal organization in the low low-level API as described [here/slide n°2](https://docs.google.com/presentation/d/1eFPfSterI6wf3CDPAKrWkp5LzjOq1RF8nC2N0YbtQMs/edit#slide=id.g338ac0a8b6_0_44).

The idea that is to offer to the consumers a client to access the clusterctl configurations as a single logical block, and then only in a second step, make explicit the fact that there are two types of configuration values.

Same applies to the provider's repositories and to the (management)cluster: 1 client to control it, and then introduce the different type of objects that they hosts.
(and BTW, in those cases, the initialization code that is deduped is more complex)

Reply:
The slide seems to indicate that the general client would apply more to the high level API.

I worry that we are quite frankly continuing to make things confusing for consumers of both the tooling and now the library by requiring a single entrypoint/interface for tasks/methods that have very different consumers (Cluster API and provider lifecycle management vs workload cluster admins).

That said, I don't consider this blocking and am willing to see how it plays out in practice.